File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;
import java.io.*;

class Main {
    static int N;
    static int[][] matrix;
    static boolean[][] visited;
    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        matrix = new int[N][N];
        
        for (int y = 0; y < N; y++) {
            char[] line = br.readLine().toCharArray();
            for (int x = 0; x < N; x++) {
                matrix[y][x] = line[x] - '0';
            }
        }
        br.close();
        
        visited = new boolean[N][N];
        List<Integer> answers = new ArrayList<>();
        for (int y = 0; y < N; y++) {
            for (int x = 0; x < N; x++) {
                if (matrix[y][x] == 1 && !visited[y][x]) {
                    answers.add(bfs(y, x));
                }
            }
        }
        
        Collections.sort(answers);
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(answers.size() + "\n");
        for(int size : answers) {
            bw.write(size + "\n");
        }
        bw.flush();
        bw.close();
    }
    
    private static int bfs(int y, int x) {
        visited[y][x] = true;
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{y, x});
        int count = 1;
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int cy = cur[0];
            int cx = cur[1];
            
            for (int dir = 0; dir < 4; dir++) {
                int ny = cy + dy[dir];
                int nx = cx + dx[dir];
                if (0 <= ny && ny < N && 0 <= nx && nx < N) {
                    if (!visited[ny][nx] && matrix[ny][nx] == 1) {
                        visited[ny][nx] = true;
                        q.add(new int[]{ny, nx});
                        count++;
                    }
                }
            }
        }
        return count;
    }
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
input = sys.stdin.readline
N = int(input())
matrix = [list(map(int, input().strip())) for _ in range(N)]
visited = [[False] * N for _ in range(N)]
answers = []

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]
def dfs(y, x):
    visited[y][x] = True
    count = 1
    
    for dir in range(4):
        ny = y + dy[dir]
        nx = x + dx[dir]
        if 0 <= ny < N and 0 <= nx < N:
            if not visited[ny][nx] and matrix[ny][nx] == 1:
                count += dfs(ny, nx)
    return count

for y in range(N):
    for x in range(N):
        if matrix[y][x] == 1 and not visited[y][x]:
            size = dfs(y, x)
            answers.append(size)

answers.sort()
print(len(answers))
print(*answers, sep = "\n")